src/net_processing.cpp

```diff
@@ -2813,15 +2817,19 @@ bool static ProcessMessage(CNode* pfrom, const std::string& strCommand, CDataStr
             // we have a chain with at least nMinimumChainWork), and we ignore
             // compact blocks with less work than our tip, it is safe to treat
             // reconstructed compact blocks as having been requested.
-            ProcessNewBlock(chainparams, pblock, /*fForceProcessing=*/true, &fNewBlock);
-            if (fNewBlock) {
+            CValidationState dos_state;
+            ProcessNewBlock(chainparams, pblock, dos_state, /*fForceProcessing=*/true, &fNewBlock);
+            if (fNewBlock && dos_state.IsValid()) {
                 pfrom->nLastBlockTime = GetTime();
             } else {
+                if (!dos_state.IsValid()) {
+                    BlockChecked(*pblock, dos_state, connman);
+                }
                 LOCK(cs_main);
                 mapBlockSource.erase(pblock->GetHash());
```

Comment:
Why is mapBlockSource.erase call needed here? In not valid case, this should already happen in explicit BlockChecked call. In valid case, I assume it happens in BlockChecked validation callback?

This logic is messy and repetitive and seems like it could be simplified by moving more of it out of here into the BlockChecked() function, or breaking BlockChecked up into reusable parts.

Reply:
This PR removes the (now redundant) BlockChecked call in case a block is completely bogus (ie !dos_state.IsValid()), so, no.

Reply:
I think you're right that we can't get rid of this `mapBlockSource.erase()` call, but not for the reason you say. We still need to call it if the block is valid, but this isn't the first time we've received it (ie `fNewBlock` is returned as true). I think this logic would be easier to follow if structured as follows, with comments:

```c++
        ProcessNewBlock(chainparams, pblock, dos_state, forceProcessing, &fNewBlock);
        if (!dos_state.IsValid()) {
            // The block failed anti-dos / mutation checks. Call BlockChecked() callback here.
            // This clears the block from mapBlockSource.
            BlockChecked(*pblock, dos_state, connman);
        } else if (!fNewBlock) {
            // Block was valid but we've seen it before. Clear it from mapBlockSource.
            LOCK(cs_main);
            mapBlockSource.erase(pblock->GetHash());
        } else {
            // Block is valid and we haven't seen it before. set nLastBlockTime for this peer.
            pfrom->nLastBlockTime = GetTime();
        }
```

Reply:
Hmm, yea, thats a bit cleaner, however, after #16323, which this was refactored out of, the original goes back to being cleaner, so would prefer to just leave it (also cause its fewer LoC changed).